spring-cloud-gcp-starters/spring-cloud-gcp-starter-core/src/main/java/org/springframework/cloud/gcp/core/GCPProperties.java

```diff
@@ -27,7 +28,7 @@
 
 	private String projectId;
 
-	private String jsonKeyFile;
+	private Resource credentialKeyFile;
```

Comment:
I don't think that a GcpProjectIdProvider makes sense, boot users are used to retrieve information from properties, that's the Spring way. Having a provider that provides a String is overkill as we mentioned here before. The Environment abstraction was created for that.

If project is a top level construct, with more information required for the utilization of the libraries (I have not seen any indication so far) than it could be something to be explored.

The idea of a auto configuration for a **Context** is to provide all the required dependencies for a cloud context to be available. `Credentials`, `Region`, `ProjectId`.

@meltsufin you would not use GcpProperties in the message handler, that's the whole point here. The message handler needs a `Credentials` object and a `String` that represents the projectId. How that is passed down will depend on how you use the library.

If you guys are still confused about this, I suggest opening a few AutoConfiguration classes from Boot, and check how the Properties from Boot play along with classes such as Redis,Rabbit fora example. The spring-data projects will require certain properties like the host and port where redis is located, they do not depend on the RedisProperties class, instead on Strings and Integers. The Autoconfiguration classes will depend on the Properties and use them to configure the connection on redis for example.